File: 02_Supervised_Learning-Regression/Python/Code/08_decision_tree_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Nov 28 20:55:17 2016

@author: pavan.govindraj
"""

#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 24 21:57:15 2016

@author: pavan.govindraj
"""


# Data preprocessing => import the dataset
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Position_Salaries.csv')

X = dataset.iloc[:, 1:2].values
Y = dataset.iloc[:, 2].values


#Splitting the dataset in to training set and test set
#No training set and test set splitting as max info is required 
'''from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size = 0.2, 
                                                    random_state = 0)
'''

#Feature scaling
"""
#May be required for some dataset 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""

#Fitting the  Decision Tree  regression model to the dataset
#create your regressor here
from sklearn.tree import DecisionTreeRegressor
regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X,Y)

#Predicting the new value from Decision Tree Regression model
y_pred = regressor.predict(6.5)

#Visualising the  regression results, since decision tree is non continuos
#use high resolution below
'''
plt.scatter(X, Y, color = 'red' )
plt.plot(X, regressor.predict(X), color = 'blue')
plt.title("Truth or Bluff( Decision Tree Regression Model)")
plt.xlabel("Position")
plt.ylabel("Salary")
plt.show()'''

#Visualising the regression results(for higher resoulution and smooth curve)
#X_grid = np.arange(min(X), max(X), 0.1)
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, Y, color = 'red' )
plt.plot(X_grid, regressor.predict(X_grid), color = 'blue')
plt.title("Truth or Bluff( Decision Tree Regression Model)")
plt.xlabel("Position")
plt.ylabel("Salary")
plt.show()

File: 02_Supervised_Learning-Regression/R/Code/08_decision_tree_regression.R
#Decision tree regression 

#Regression template

#Data Preprocessing ; importing the

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

#Splitting the dataset in to training set and test set
#install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 0.8 )
# training_set = subset(dataset, split == TRUE)
# test_set= subset(dataset, split == FALSE)

#Feature scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

#Fitting the decision tree regression model to the dataset
#create a regressor
#install.packages('rpart')
library('rpart')
regressor = rpart(formula = Salary ~ .,
                  data = dataset,
                  control = rpart.control(minsplit = 1))
#Predict a new result with decision tree regression model
y_pred = predict(regressor, newdata = data.frame(Level = 6.5))

#Visualising the decision tree regression model results, since decision tree is non-continuos
#we need to use high resolution below 
# library(ggplot2)
# ggplot() +
#   geom_point(aes(x = dataset$Level, y = dataset$Salary),
#              color = 'red') +
#   geom_line(aes(x = dataset$Level, y = predict(regressor, newdata = dataset)),
#             color = 'blue') +
#   ggtitle("Truth or Bluff(Decision Tree Regression Model)") +
#   xlab("Level") +
#   ylab("Salary")

#Visualising the decision tree regression model results
#For high resolution and smoothen the curve
library(ggplot2)
#X_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
X_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = X_grid, y = predict(regressor, newdata = data.frame(Level = X_grid))),
            color = 'blue') +
  ggtitle("Truth or Bluff(Decision Tree Regression Model)") +
  xlab("Level") +
  ylab("Salary")
